vivid-server/src/modules/friends/friends.controller.ts

```diff
@@ -31,57 +31,114 @@ export class FriendsController {
     return this.friendsService.findAll();
   }
 
+  // Send friend request
   @Post('add/:friend_id')
   async friendRequest(
     @Param('friend_id') friendId: string,
     @User() user: UserEntity,
-    @Res() response: Response,
+    @Req() req: Request,
   ) {
     let friend = await this.userService.findUser(friendId);
 
     // checking if friend is in general user table
     if (!friend) {
-      response.status(404).send('user not found');
-      throw 'error';
+      req.res.status(404).send('user not found');
+      return;
     }
-    
+
     // checking if friend is the logged in user
     if (user.id === friend.id) {
-      response.status(404).send("can't be friends with ya self mate");
-      throw 'error';
+      req.res.status(404).send("can't be friends with ya self mate");
+      return;
+    }
+
+    //checking if friend request has already been made or if they're already friends
+    if (
+      this.friendsService.findFriendRequest(user.id, friend.id) ||
+      this.friendsService.findFriendRequest(friend.id, user.id) ||
+      this.friendsService.findFriend(friend.id, user.id)
+    ) {
```

Comment:
possible race conditions. fix with:
1. Constraints make user1 and user2 a guaranteed sort [see this](https://github.com/typeorm/typeorm/issues/701#issuecomment-374681109)
2. UNIQUE contraint for user1 and user2. so the combination of them is unique, [see this](https://github.com/typeorm/typeorm/issues/1616#issuecomment-657022282)